File: core/src/main/java/dev/kkorolyov/pancake/core/system/AudioSystem.java
package dev.kkorolyov.pancake.core.system;

import dev.kkorolyov.pancake.core.component.AudioEmitter;
import dev.kkorolyov.pancake.core.component.Transform;
import dev.kkorolyov.pancake.platform.Config;
import dev.kkorolyov.pancake.platform.plugin.GameSystem;
import dev.kkorolyov.pancake.platform.entity.Entity;
import dev.kkorolyov.pancake.platform.entity.Signature;
import dev.kkorolyov.pancake.platform.math.Vector3;
import dev.kkorolyov.pancake.platform.math.VectorMath;
import dev.kkorolyov.pancake.platform.math.Vectors;
import dev.kkorolyov.pancake.platform.plugin.Plugins;
import dev.kkorolyov.pancake.platform.plugin.RenderMedium;
import dev.kkorolyov.pancake.platform.utility.Limiter;

/**
 * Starts and stops audio clips.
 */
public class AudioSystem extends GameSystem {
	private final RenderMedium renderMedium;

	private final Vector3 relativeEmitter = Vectors.create(0, 0, 0);

	/**
	 * Constructs a new sound system.
	 */
	public AudioSystem() {
		this(Plugins.renderMedium());
	}
	AudioSystem(RenderMedium renderMedium) {
		super(
				new Signature(AudioEmitter.class, Transform.class),
				Limiter.fromConfig(AudioSystem.class)
		);
		this.renderMedium = renderMedium;
	}

	@Override
	public void update(Entity entity, long dt) {
		AudioEmitter emitter = entity.get(AudioEmitter.class);
		Vector3 position = entity.get(Transform.class).getGlobalPosition();

		relativeEmitter.set(position);
		relativeEmitter.add(renderMedium.getCamera().getPosition(), -1);

		double centralRadius = Double.parseDouble(Config.get(getClass()).getProperty("centralRadius"));

		emitter.apply(
				centralRadius / VectorMath.magnitude(relativeEmitter),
				relativeEmitter.getX() / centralRadius
		);
	}
}

File: audio-jfx/src/main/kotlin/dev/kkorolyov/pancake/audio/jfx/system/AudioSystem.kt
package dev.kkorolyov.pancake.audio.jfx.system

import dev.kkorolyov.pancake.audio.jfx.ReceiverQueue
import dev.kkorolyov.pancake.audio.jfx.component.AudioEmitter
import dev.kkorolyov.pancake.core.component.Transform
import dev.kkorolyov.pancake.platform.Config
import dev.kkorolyov.pancake.platform.GameSystem
import dev.kkorolyov.pancake.platform.entity.Entity
import dev.kkorolyov.pancake.platform.math.Vector3
import dev.kkorolyov.pancake.platform.math.Vectors
import kotlin.math.max
import kotlin.math.min

/**
 * Maintains audio state relative to the current receiver.
 */
class AudioSystem(private val queue: ReceiverQueue) : GameSystem(AudioEmitter::class.java, Transform::class.java) {
	private var active = true

	private val emitPoint = Vectors.create3()

	override fun update(entity: Entity, dt: Long) {
		val emitter = entity.get(AudioEmitter::class.java)
		val transform = entity.get(Transform::class.java)

		emitter.active = active

		var volume = 1.0
		var balance = 0.0

		queue.position?.let {
			emitPoint.set(transform.position)
			emitPoint.add(it, -1.0)

			// reduce emitter position sensitivity by expanding a point to a sphere of some radius
			val audioRadius = Config.get(javaClass).getProperty("audioRadius").toDouble()

			volume = audioRadius / Vector3.magnitude(emitPoint)
			balance = emitPoint.x / audioRadius
		}

		emitter(
			queue.volume * min(1.0, volume),
			if (balance < 0) max(-1.0, balance) else min(1.0, balance)
		)
	}
}
